File: common/src/main/java/com/ssblur/scriptor/helpers/resource/SpellResource.java
package com.ssblur.scriptor.helpers.resource;

import com.google.gson.JsonPrimitive;

import java.util.List;

public class SpellResource {
  public static class PartialSpellResource {
    public String action;
    public List<JsonPrimitive> descriptors;

    public PartialSpellResource(String action, List<JsonPrimitive> descriptors) {
      this.action = action;
      this.descriptors = descriptors;
    }
  }

  public String subject;
  public List<PartialSpellResource> spells;
  public SpellResource(String subject, List<PartialSpellResource> spells) {
    this.subject = subject;
    this.spells = spells;
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/resources/shared/SpellResource.kt
package com.ssblur.scriptor.resources.shared

import com.ssblur.scriptor.error.WordNotFoundException
import com.ssblur.scriptor.registry.words.WordRegistry.actionRegistry
import com.ssblur.scriptor.registry.words.WordRegistry.descriptorRegistry
import com.ssblur.scriptor.registry.words.WordRegistry.subjectRegistry
import com.ssblur.scriptor.word.PartialSpell
import com.ssblur.scriptor.word.Spell

class SpellResource(var subject: String, var spells: List<PartialSpellResource>, var spellData: List<String>?) {
  class PartialSpellResource(var action: String, var descriptors: List<String>)
  fun getSpell(): Spell {
    val spells = mutableListOf<PartialSpell>()
    for (spell in this.spells) {
      val action = actionRegistry[spell.action] ?: throw WordNotFoundException(spell.action)
      val descriptors = spell.descriptors.map { descriptorRegistry[it] ?: throw WordNotFoundException(it) }
      spells += PartialSpell(action, *descriptors.toTypedArray())
    }
    return Spell(
      subjectRegistry[this.subject] ?: throw WordNotFoundException(this.subject),
      *spells.toTypedArray(),
      spellData = spellData ?: listOf()
    )
  }
}